File: Ex56RecyclerView/app/src/main/java/com/jmnl2020/ex56recyclerview/MyAdapter.java
package com.jmnl2020.ex56recyclerview;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class MyAdapter extends RecyclerView.Adapter {

    Context context;
    ArrayList<String> items;


    //1. constructor 생성

    public MyAdapter(Context context, ArrayList<String> items) {
        this.context = context;
        this.items = items;
    }


    //재활용할 뷰가 없으면 뷰를 만들어내는 작업 수행하는 메소드
    //단, 뷰홀더 객체를 만들어서 리턴!
    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(context);
        View itemView = inflater.inflate(R.layout.recycler_item, parent, false);

        // 2. 맨 아래에 뷰홀더 객체 생성, 참조변수 대입
        VH holder = new VH(itemView);

        return holder;
    }

    // 뷰에 값을 연결하는 작업메소드
    // 뷰에 설정된 tag객체 (ViewHolder)를 첫번째 파라미터로 전달
    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {

        //3. vh 홀더 대입
        VH vh= (VH) holder;
        vh.tv.setText(items.get(position));

    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    //이너클래스 : 아이템뷰 안의 뷰들의 참조변수를 멤버변수로 가질 클래스
    //=> 아이템뷰 안의 뷰에 아이디를 찾아주기 위해 생성
    class VH extends RecyclerView.ViewHolder {

        TextView tv;

        public VH(@NonNull View itemView) {
            super(itemView);

            tv= itemView.findViewById(R.id.tv);

            //리사이클러의 아이템클릭에 반응하는 리스너객체 생성 및 추가
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    //이 아이템뷰의 위치(순서) index num
                    int position = getLayoutPosition();

                    Toast.makeText(context, ""+position, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }

}

File: Ex91KoptlinRecyclerViewApp/app/src/main/java/com/jmnl2020/ex91koptlinrecyclerviewapp/MyAdapter.kt
package com.jmnl2020.ex91koptlinrecyclerviewapp

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.recycler_item.view.*

class MyAdapter constructor(val context:Context, val items: ArrayList<ItemVO>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val layoutInflater = LayoutInflater.from(context)
        val itemView = layoutInflater.inflate(R.layout.recycler_item, parent, false)
        val vh= VH(itemView)
        return vh
    }

    override fun getItemCount(): Int {
        return items.size
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val vh = holder as VH // 코틀린에서 형변환 하는 방법 : 연산자 as 를 사용

        //현재번째 아이템
        val item = items.get(position)
        vh.itemView.tvTitle.setText(item.title)
        //vh.itemView.tvMsg.setText(item.msg)
        //코틀린에서는 setXXX메소드를 싫어함.
        //멤버변수(property:속성)로 설정하는 것을 권장
        vh.itemView.tvMsg.setText(item.mgs)


        //glide 와 같은 역할을 하는 Picasso 라이브러리
        Picasso.get().load(item.img).into(vh.itemView.iv)

        //코틀린에서는 이 위치에서 itemView의 클릭이벤트 처리를 함
//        vh.itemView.setOnClickListener(object : View.OnClickListener{
//            override fun onClick(v: View?) {
//               //아이템 상세 정보화면으로 이동
//            }
//        })

    }

    //inner class ViewHolder 클래스 : itemView안의 뷰들을 관리하는 클래스
    inner class VH constructor(itemView: View) : RecyclerView.ViewHolder(itemView){

//        val tvTitle = itemView.tvTitle
//        val tvMsg = itemView.tvMsg
//        val iv = itemView.iv

        init {
            //자바에ㅓㅅ는 원래 이 생성자에서 getLayoutPosition() 메소드로
            //클릭한 아이템을 구별했음
            // 코틀립에서는 getLayourPosition()메소드 대신에
            //이 아답터에 멤버변수로 이미 layoutPositioni이 존재함
            //근데 layoutPutposition() 속성값이 언제나 항상 -1 !!!
            //그래서 이 곳에서 클릭 위치를 찾을수가 없다.

            itemView.setOnClickListener(object :View.OnClickListener{
                override fun onClick(v: View?) {
                    //여기서 상세 아이템 페이지

                    val intent = Intent(context, ItemActivity::class.java)
                    intent.putExtra("title", items.get(layoutPosition).title)
                    intent.putExtra("msg", items.get(layoutPosition).mgs)
                    intent.putExtra("img",items.get(layoutPosition).img)

                    context.startActivity(intent)

                }
            })



        }



    }



}
